### PR TITLE
[LottieViewer] Pixel View feature added

### PR DESCRIPTION
**Pull request by aborziak-ms**
Adding **PixelView** feature. This feature should allow designers to examine animations precisely pixel-by-pixel, so that they can ensure that animation will be displayed correctly at any resolution.

PixelView

### CODE REVIEW

File: LottieViewer/PixelViewElement.cs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+using Windows.UI.Xaml;
+using Windows.UI.Xaml.Controls;
+using Windows.UI.Xaml.Hosting;
+using Windows.UI.Xaml.Input;
+
+#nullable enable
+
+namespace LottieViewer
+{
+    public sealed class PixelViewElement : UserControl
```
Comment (eliezerpMS) [inline LottieViewer/PixelViewElement.cs:22]: >PixelViewElement [](http://example.com/codeflow?start=24&length=16)

Nit: put a comment summarizing what this class does.

File: LottieViewer/PixelViewElement.cs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            var c = ElementCompositionPreview.GetElementVisual(this).Compositor;
+            _visual = c.CreateSpriteVisual();
+            _visual.RelativeSizeAdjustment = Vector2.One;
+            ElementCompositionPreview.SetElementChildVisual(this, _visual);
+        }
+
+        public void OnMouseMove(object sender, PointerRoutedEventArgs e)
+        {
+            var position = e.GetCurrentPoint(this).Position;
+            if (_previousScale < 0 || _previousBitmap == null || ActualHeight <= 0)
```
Comment (eliezerpMS) [inline LottieViewer/PixelViewElement.cs:77]: >_previousScale [](http://example.com/codeflow?start=16&length=14)

Can this ever be less than 0 can we make this a uint?

File: LottieViewer/PixelViewElement.cs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+        public Vector2 CurrentPosition
+        {
+            get { return (Vector2)GetValue(CurrentPositionProperty); }
+            set { SetValue(CurrentPositionProperty, value); }
+        }
+
+        public PixelViewElement()
+        {
+            var c = ElementCompositionPreview.GetElementVisual(this).Compositor;
```
Comment (eliezerpMS) [inline LottieViewer/PixelViewElement.cs:68]: >c [](http://example.com/codeflow?start=16&length=1)

nit:  compositor

File: LottieViewer/PixelViewElement.cs (hunk 2 above)
Comment (aborziak-ms) [reply LottieViewer/PixelViewElement.cs:77]: you are right, it is not used anymore, I will delete

File: LottieViewer/MainPage.xaml.cs [hunk 4]
```diff
@@ -60,6 +64,16 @@ public MainPage()
 
             // Remove all of the control panel panes. They will be added back as needed.
             ControlPanel.Children.Clear();
+
+            timer = new Timer(TimerCallback, null, 0, (int)TimeSpan.FromMilliseconds(200).TotalMilliseconds);
+        }
+
+        private void TimerCallback(object state)
+        {
+            _ = Dispatcher.RunAsync(CoreDispatcherPriority.High, () =>
```
Comment (eliezerpMS) [inline LottieViewer/MainPage.xaml.cs:73]: >High [](http://example.com/codeflow?start=59&length=4)

Why high? (I feel like i should know this but at least for the documentation it recommends to use normal)

File: LottieViewer/PixelViewElement.cs [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            }
+
+            _currentColor = _previousColors[(pixelY * _previousBitmap.SizeInPixels.Width) + pixelX];
+            SetValue(CurrentColorStringProperty, _currentColor.ToString());
+            SetValue(CurrentPositionProperty, new Vector2(pixelX, pixelY));
+        }
+
+        public async Task UpdatePixelViewAsync(FrameworkElement element)
+        {
+            _comparisonIsQueued = true;
```
Comment (eliezerpMS) [inline LottieViewer/PixelViewElement.cs:101]: >_comparisonIsQueued [](http://example.com/codeflow?start=12&length=19)

Comparison? Should this be called something else?

File: LottieViewer/MainPage.xaml.cs [hunk 6]
```diff
@@ -60,6 +64,16 @@ public MainPage()
 
             // Remove all of the control panel panes. They will be added back as needed.
             ControlPanel.Children.Clear();
+
+            timer = new Timer(TimerCallback, null, 0, (int)TimeSpan.FromMilliseconds(200).TotalMilliseconds);
```
Comment (eliezerpMS) [inline LottieViewer/MainPage.xaml.cs:68]: >TimerCallback [](http://example.com/codeflow?start=30&length=13)

Given this part I am wondering if DispatcherTimer is more appropriate?

Comment (eliezerpMS) [inline LottieViewer/MainPage.xaml.cs:68]: >200 [](http://example.com/codeflow?start=85&length=3)

Just curious why 200?

File: LottieViewer/PixelViewElement.cs [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+                    session.StartCapture();
+
+                    // Wait for the frame to arrive.
+                    var result = await tcs.Task;
+
+                    // !!!!!!!! NOTE !!!!!!!!
+                    // This thread is now running inside the OnFrameArrived callback method.
+
+                    // Unsubscribe now that we have captured the frame.
+                    framePool.FrameArrived -= OnFrameArrived;
```
Comment (eliezerpMS) [inline LottieViewer/PixelViewElement.cs:222]: >framePool.FrameArrived -= OnFrameArrived; [](http://example.com/codeflow?start=20&length=41)

Is there an advantage to subscribing and unsubscribing? 

Also is there a reason to create and destroy things like the framepool? 

Wondering about the case where we call this with a repeat timer.

File: LottieViewer/PixelViewElement.cs [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+                using (var session = framePool.CreateCaptureSession(graphicsItem))
+                {
+                    framePool.FrameArrived += OnFrameArrived;
+
+                    // Start capturing. The FrameArrived event will occur shortly.
+                    session.StartCapture();
+
+                    // Wait for the frame to arrive.
+                    var result = await tcs.Task;
```
Comment (eliezerpMS) [inline LottieViewer/PixelViewElement.cs:216]: >var result [](http://example.com/codeflow?start=20&length=10)

For consistency can this be "_"? We really care about the OnFrameArrived continuation right?

**Review by eliezerpMS — APPROVED** [review]
:shipit:

**Pull request merged**